pkg/counter/cache.go

```diff
@@ -0,0 +1,162 @@
+package counter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var cachedCounts = cache{
+	Counts: Counts{
+		UserAccountsPerClusterCounts: map[string]int{},
+	},
+}
+
+// Counts is type that contains number of MURs and number of UserAccounts per member cluster
+type Counts struct {
+	MasterUserRecordCount        int
+	UserAccountsPerClusterCounts map[string]int
+}
+
+type cache struct {
+	Counts
+	sync.RWMutex
+	initialized bool
+}
+
+func write(operation func()) {
+	cachedCounts.Lock()
+	defer cachedCounts.Unlock()
+	operation()
+}
+
+// Reset resets the cached counter - is supposed to be used only in tests
+func Reset() {
+	write(func() {
+		reset()
+	})
+}
+
+func reset() {
+	cachedCounts.Counts = Counts{
+		UserAccountsPerClusterCounts: map[string]int{},
+	}
+	cachedCounts.initialized = false
+}
+
+// IncrementMasterUserRecordCount increments the number of MasterUserRecord in the cached counter
+func IncrementMasterUserRecordCount() {
+	write(func() {
+		cachedCounts.MasterUserRecordCount++
+	})
+}
+
+// DecrementMasterUserRecordCount decreases the number of MasterUserRecord in the cached counter
+func DecrementMasterUserRecordCount(log logr.Logger) {
+	write(func() {
+		if cachedCounts.MasterUserRecordCount != 0 {
+			cachedCounts.MasterUserRecordCount--
+		} else {
+			log.Error(fmt.Errorf("the count of MasterUserRecords is zero"),
+				"unable to decrement the number of MasterUserRecords")
+		}
+	})
+}
+
+// IncrementUserAccountCount increments the number of UserAccount fo the given member cluster in the cached counter
+func IncrementUserAccountCount(clusterName string) {
+	write(func() {
+		cachedCounts.UserAccountsPerClusterCounts[clusterName]++
+	})
+}
+
+// DecrementUserAccountCount decreases the number of UserAccount fo the given member cluster in the cached counter
+func DecrementUserAccountCount(log logr.Logger, clusterName string) {
+	write(func() {
+		if cachedCounts.UserAccountsPerClusterCounts[clusterName] != 0 {
+			cachedCounts.UserAccountsPerClusterCounts[clusterName]--
+		} else {
+			log.Error(fmt.Errorf("the count of UserAccounts is zero"),
+				"unable to decrement the number of UserAccounts for the given cluster", "cluster", clusterName)
+		}
+	})
+}
+
+// GetCounts returns Counts struct containing number of MURs and number of UserAccounts per member cluster.
+// If the counter is not yet initialized, then it returns error
+func GetCounts() (Counts, error) {
+	cachedCounts.RLock()
+	defer cachedCounts.RUnlock()
+	if !cachedCounts.initialized {
+		return cachedCounts.Counts, fmt.Errorf("counter is not initialized")
```

Comment:
This is slightly tricky - we need to initialize with the existing `ToolchainStatus`. I should not retrieve the `ToolchainStatus` before having the client cache synced, but waiting for the cache being synced can be slower than the execution in controllers. Thus, I decided to let the `ToolchainStatus` initialize the cache during the first reconcile, which should be immediate.
Theoretically, when the `UserSignup` controller will read the cache, it will probably also have the `ToolchainStatus` retrieved, so it will be possible to pass it together with the resource. On the other hand, the only entity who should update and synchronize the `ToolchainStatus` resource is its controller and not `UserSignup`. So, I would really let the `ToolchainStatus` initialize the cache - there won't be almost not problem with it - in the worst case, the controller will rerun the reconciliation for unapproved `UserSignups` based on the returned error one more time.